### PR TITLE
[Feature] Add LTX-2.3 keyframe interpolation (FLF2V / FMLF) one-stage pipeline

### PR DESCRIPTION
**Pull request by lucasbarrez**
## What

Adds `LTX23ConditionPipeline`, a one-stage multi-anchor **keyframe-interpolation**
pipeline for LTX-2.3 (FLF2V / FMLF), built on the unified `LTXPipelineRuntime`
from #5147. Anchors pin arbitrary frame indices (each with its own strength) and
the model interpolates the motion between them under text + audio conditioning.

Addresses the *keyframe interpolation* item of #4985.

## How

- **`LTXMultiAnchorConditioningMixin`** — sibling to `LTXI2VConditioningMixin`.
  Where I2V pins only frame 0, this generalizes to a set of anchor indices,
  reusing the existing `conditioning_mask` / video-audio-scheduler / timestep
  hooks. **No shared runtime file is modified.**
- **`LTX23ConditionPipeline(LTXMultiAnchorConditioningMixin, LTX23Pipeline)`** —
  thin MRO composition; inherits x0 guidance, component profile and denoise loop.
  Degrades to pure T2V when no conditions are supplied.
- **`ConditionVideoAudioScheduler`** — applies the x0-space anchor blend during
  the loop; the mask and clean latents are captured at construction, so no
  mutable per-step pipeline state is introduced.
- Everything lives in a self-contained new module `ltx2_condition.py`.

## Files

- `vllm_omni/diffusion/models/ltx2/ltx2_condition.py` (new)
- `vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py`, `__init__.py`,
  `vllm_omni/diffusion/registry.py` (register + export)
- `tests/diffusion/models/ltx2/test_ltx2_condition.py` (new)
- `docs/models/supported_models.md`

## Testing

- **Unit + structural (CI, `core_model`/`cpu`, no GPU):** MRO / guidance / hook /
  registry contracts, condition validation, `apply_visual_conditioning` token
  pinning.
- **Forward-path (CI, CPU):** condition resolution from `multi_modal_data`,
  pure-T2V fallback when no conditions, and the anchor timestep-masking hook
  (`timestep = t·(1 − mask)`, audio unmasked).
- **End-to-end (manual, H100 / Modal):** FLF2V & FMLF requests generate coherent
  video + audio with anchors correctly pinned at first / middle / last frames.
- **Deferred to follow-up:** a checked-in GPU accuracy/similarity E2E test (needs
  weights + GPU CI), in the spirit of `tests/e2e/accuracy/test_ltx2_3_video_similarity.py`.

## Scope / follow-ups

- **One-stage only.** The two-stage condition variant is intentionally held until
  the Standard / Full-distilled recipe layer lands, so it builds on that rather
  than the legacy `LTX2TwoStagesPipeline` path.
- **HTTP serving** for `conditions` (request schema → `multi_modal_data["conditions"]`)
  is a **follow-up PR**. The pipeline reads conditions defensively, so it is
  independently mergeable. Two open contract questions for the serving layer:
  1. Should multi-anchor input flow through the same `multi_modal_data` path as
     I2V, or do you want a dedicated anchor schema?
  2. Any objection to a public `LTX23ConditionPipeline` class as the entry point,
     mirroring `LTX23ImageToVideoPipeline`?

---

cc @mglyn — following up on #4985, where you noted keyframe interpolation would be
welcome once it can build on the shared pipeline framework. This is that work,
re-expressed on the merged `LTXPipelineRuntime`. Happy to adjust the entry-point
and request-schema contract per your answer to the two questions above.

### CODE REVIEW

File: vllm_omni/diffusion/models/ltx2/ltx2_condition.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+    """Overwrite ``latents`` / ``conditioning_mask`` at the conditioned token
+    indices and build the parallel ``clean_latents`` tensor.
+
+    Returns ``(latents, conditioning_mask, clean_latents)``. Matches Diffusers'
+    ``LTX2ConditionPipeline.apply_visual_conditioning``.
+    """
+    clean_latents = torch.zeros_like(latents)
+    for cond, strength, latent_idx in zip(condition_latents, condition_strengths, condition_indices, strict=True):
+        num_cond_tokens = cond.size(1)
+        start_token_idx = latent_idx * latent_height * latent_width
```
Comment (david6666666) [inline vllm_omni/diffusion/models/ltx2/ltx2_condition.py:177]: Non-first anchors cannot replace the base latent timeline this way. The current Diffusers LTX-2.3 condition pipeline appends every non-zero-index anchor as extra keyframe tokens with its own positional coordinates; `apply_visual_conditioning()` only overwrites index 0. Here a middle/last anchor gets base-frame RoPE positions, so FLF/FMLF conditioning is not equivalent. Please port the keyframe-token/coordinate path and add a non-zero-index regression test.

File: vllm_omni/diffusion/models/ltx2/ltx2_condition.py [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+        """Read condition anchors from the request's ``multi_modal_data``.
+
+        The OpenAI-compatible serving layer stashes a list of objects exposing
+        ``.data`` (PIL image), ``.index`` (int), and ``.strength`` (float) under
+        ``prompt["multi_modal_data"]["conditions"]``. Duck-typed here so the
+        pipeline does not depend on serving types.
+        """
+        if not req.prompts:
+            return None
+        first = req.prompts[0]
```
Comment (david6666666) [inline vllm_omni/diffusion/models/ltx2/ltx2_condition.py:229]: `LTX23Pipeline` supports request batching, but this resolver reads only `req.prompts[0]` and `_prepare_condition_latents()` then applies that condition set to the whole batch. A batched request will therefore use the first request's anchors for every item. Please resolve anchors per request (or disable batching for this pipeline) and cover mixed condition sets.

**Comment by lucasbarrez** [comment]
Thanks @david6666666 — you're right, and good catch. I'd initially matched **diffusers 0.38.0** (the version vLLM-Omni pins), where `apply_visual_conditioning` overwrites *all* anchors in place. **diffusers `main` changed this**: non-first-frame anchors are appended as keyframe tokens with their own coordinates (`VideoConditionByKeyframeIndex`), which correctly positions a middle/last anchor at its own pixel-frame index instead of a base-timeline slot.

Ported that behavior in 0e0271d5:

- **`apply_first_frame_conditioning()`** only overwrites `latent_idx == 0` now.
- **`_prepare_keyframe_coords()`** computes per-keyframe pixel-space RoPE coordinates (no causal fix, offset by `pixel_frame_idx`, single-frame temporal clamp) — a direct port of `_prepare_keyframe_coords` from diffusers main.
- Keyframe **tokens/coords/mask are appended** to the sequence in `_prepare_denoise_context_for_cfg` (before the CFG duplication) and **trimmed** in `_unpack_and_denormalize_stage`. Both are overridable pipeline hooks, so **no shared runtime file is touched** — the transformer already drives RoPE entirely from `video_coords`, so extending latents + coords together is enough.
- Added a **non-zero-index regression test** (`_prepare_keyframe_coords` positioning + `apply_first_frame_conditioning` skipping non-zero indices).

Validated FLF2V / FMLF end-to-end on H100. Thanks for the thorough review!

**Comment by lucasbarrez** [comment]
Heads up — I see the LTX unification landed (`LTX2Pipeline` / `LTXRuntime`, version resolved from checkpoint metadata). Rebasing this onto the new structure now and re-expressing the condition pipeline on `LTX2Pipeline`; will force-push the updated branch shortly. The keyframe-token logic itself is unchanged, just the base class / hook plumbing.

**Comment by david6666666** [comment]
@mglyn ptal thx

File: vllm_omni/diffusion/models/ltx2/ltx2_condition.py (hunk 1 above)
Comment (david6666666) [reply vllm_omni/diffusion/models/ltx2/ltx2_condition.py:177]: The rebased keyframe-token/coordinate path and focused regression test address this. Looks good now.

File: vllm_omni/diffusion/models/ltx2/ltx2_condition.py (hunk 2 above)
Comment (david6666666) [reply vllm_omni/diffusion/models/ltx2/ltx2_condition.py:229]: Disabling request batching for this pipeline addresses this. Looks good now.

File: vllm_omni/entrypoints/openai/api_server.py [hunk 3]
```diff
@@ -2950,13 +2953,15 @@ async def _parse_video_form(
     frame_interpolation_model_path: str | None = Form(default=None),
     lora: str | None = Form(default=None),
     extra_params: str | None = Form(default=None),
+    conditions: str | None = Form(default=None),
```
Comment (david6666666) [inline vllm_omni/entrypoints/openai/api_server.py:2956]: This adds a public multipart API path, but the only new tests cover the pipeline directly. Please add endpoint coverage for decoding and forwarding anchors, plus a 400 response for invalid URLs or indices, for both sync and async video routes.

File: vllm_omni/diffusion/models/ltx2/ltx2_condition.py [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+  tokens* with their own RoPE coordinates (``_prepare_keyframe_coords``), so a
+  middle/last anchor is positioned at its own pixel-frame index. The sequence
+  grows by those tokens for the denoise loop and is trimmed before VAE decode.
+
+Re-expressed on the unified ``LTXRuntime``: conditioning is a mixin over
+``LTX2Pipeline`` mirroring ``LTXI2VConditioningMixin``. The anchor pinning runs
+through the ``_step_video_latents_i2v`` hook (which the shared step adapter calls
+whenever a conditioning mask is present), so no shared runtime file is modified.
+"""
+
```
Comment (mglyn) [inline vllm_omni/diffusion/models/ltx2/ltx2_condition.py:21]: #5148 already introduced the phase-aware LTXI2VConditioningMixin in
ltx2_conditioning.py, shared by one- and two-stage entries. Generalize that
visual-conditioning path into first-frame, combined, and guiding-keyframe
policies instead of duplicating preprocessing, VAE, mask, timestep/guidance,
and Euler-conditioning hooks in a parallel module.

File: vllm_omni/diffusion/models/ltx2/ltx2_condition.py [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+    pixel_frames: list[int] = []
+    for cond in conditions:
+        if not 0.0 <= cond.strength <= 1.0:
+            raise ValueError(f"Condition strength must be in [0, 1], got {cond.strength}.")
+        resolved = cond.index if cond.index >= 0 else latent_num_frames + cond.index
+        if not 0 <= resolved < latent_num_frames:
+            raise ValueError(
+                f"Condition index {cond.index} resolves to {resolved}, "
+                f"outside the valid range [0, {latent_num_frames})."
+            )
```
Comment (mglyn) [inline vllm_omni/diffusion/models/ltx2/ltx2_condition.py:110]: This differs from the current Diffusers/LTX reference preprocessing:
aspect-fill, center crop, F.interpolate, and duration-aware trimming.
VideoProcessor stretches the aspect ratio and leaves overlong clips intact.

File: vllm_omni/diffusion/models/ltx2/ltx2_condition.py [hunk 6]
```diff
@@ -0,0 +1,589 @@
...
+    latent_height: int,
+    latent_width: int,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    """Overwrite ``latents`` / ``conditioning_mask`` in place at the first-frame
+    positions and build the parallel ``clean_latents`` tensor.
+
+    Only conditions with ``latent_idx == 0`` are applied here (matching Diffusers'
+    ``apply_first_frame_conditioning``); non-first-frame anchors are appended as
+    keyframe tokens instead and are skipped. ``conditioning_mask`` is ``(B, N)``.
+    """
```
Comment (mglyn) [inline vllm_omni/diffusion/models/ltx2/ltx2_condition.py:171]: Replacing frame 0 and appending only non-zero anchors matches Diffusers
and combined_image_conditionings, not the specialized official keyframe
pipeline, which appends frame 0 as a guiding latent as well.

File: vllm_omni/diffusion/models/ltx2/ltx2_condition.py [hunk 7]
```diff
@@ -0,0 +1,589 @@
...
+                mask=torch.cat(kf_mask, dim=1),
+            )
+
+        noise = randn_tensor(latents.shape, generator=generator, device=latents.device, dtype=latents.dtype)
+        effective_scale = noise_scale if latents_supplied else 1.0
+        base_scaled = (effective_scale * (1.0 - conditioning_mask)).unsqueeze(-1)
+        latents = base_scaled * noise + (1.0 - base_scaled) * latents
+
+        # Renoise keyframe tokens with the same formula (strength=1 keeps them clean).
+        if keyframe is not None:
```
Comment (mglyn) [inline vllm_omni/diffusion/models/ltx2/ltx2_condition.py:407]: Preserve a separate clean copy before noising. The current x0
blend targets the initially noised keyframe for strength < 1;
Diffusers keeps a separate kf_clean_list as the clean target.

File: vllm_omni/diffusion/models/ltx2/ltx2_condition.py [hunk 8]
```diff
@@ -0,0 +1,589 @@
...
+        device: torch.device,
+        noise_scale: float,
+        image: Any | None = None,
+    ) -> tuple[torch.Tensor, torch.Tensor | None]:
+        conditions = getattr(self, "_pending_conditions", None)
+        if not conditions:
+            return super()._prepare_video_latents_stage(
+                request_inputs, prompt_context, device=device, noise_scale=noise_scale, image=image
+            )
+
```
Comment (mglyn) [inline vllm_omni/diffusion/models/ltx2/ltx2_condition.py:430]: num_videos_per_prompt can make B > 1 while every keyframe tensor has
B=1, so sequence-dimension concatenation fails. Repeat anchors across
the output batch or explicitly reject multiple outputs.

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py [hunk 9]
```diff
@@ -36,6 +37,15 @@ class LTX2Pipeline(LTXI2VConditioningMixin, LTXRuntime):
     supports_request_batch = True
```
Comment (mglyn) [inline vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py:39]: This is a Diffusers-style one-stage pipeline. The official Lightricks
pipeline is two-stage, re-applies keyframes after upsampling, and uses the
distilled LoRA in Stage 2; do not claim full official parity yet.

File: vllm_omni/entrypoints/openai/api_server.py
Comment (mglyn) [inline vllm_omni/entrypoints/openai/api_server.py]: This keyword is passed even when None, but not every existing
handler/test double accepts it; test_video_server.py currently fails
with BlockingVideoHandler.generate_video_bytes() TypeError.

File: vllm_omni/entrypoints/openai/serving_video.py [hunk 10]
```diff
@@ -140,13 +154,27 @@ async def _run_and_extract(
             target_size = (vp.width, vp.height)
             if input_image.size != target_size:
                 input_image = input_image.resize(target_size, Image.Resampling.LANCZOS)
```
Comment (mglyn) [inline vllm_omni/entrypoints/openai/serving_video.py:156]: Online uses PIL/LANCZOS while
offline uses VideoProcessor/bilinear, producing different, distorted
inputs. Normalize once in the pipeline with reference preprocessing.

File: vllm_omni/entrypoints/openai/api_server.py
Comment (mglyn) [inline vllm_omni/entrypoints/openai/api_server.py:2853]: This keyword is passed even when None, but not every existing
handler/test double accepts it; test_video_server.py currently fails
with BlockingVideoHandler.generate_video_bytes() TypeError.

File: vllm_omni/entrypoints/openai/api_server.py [hunk 11]
```diff
@@ -3068,7 +3074,29 @@ async def _parse_video_form(
             raise HTTPException(400, detail=str(exc)) from exc
         reference_audio = ReferenceAudio(path=audio_path)
 
-    return request, handler, effective_model_name, reference_image, reference_video, reference_audio
```
Comment (mglyn) [inline vllm_omni/entrypoints/openai/api_server.py:3071]: Validate pipeline capability before decoding anchors and reject mixing
conditions with ordinary references. Otherwise conditions, or the ordinary
image in LTX2ConditionPipeline, can be silently ignored.

**Comment by lucasbarrez** [comment]
Thanks so much @mglyn and @david6666666 for the thorough and thoughtful review, I really appreciate the time you took digging into the details. 🙏

@david6666666, glad the keyframe token/coordinate rework and the batching change addressed your points, and good call on the endpoint tests.

@mglyn, this all makes complete sense and I'm fully on board with the direction. A shared LTX conditioning framework (generic latent conditioning, masks/positions, per-phase conditions, reference tokens, one-/two-stage reuse) is clearly the right foundation, much cleaner than a parallel pipeline, especially with two-stage, retake, IC-LoRA, A2V and LipDub all on the way. Happy to keep this PR open and re-express the work as a small integration + tests on top of the new interfaces once they land. I'll also fix the test_video_server.py regression I introduced in the meantime. 

Really looking forward to the new infrastructure, thanks for letting me continue contributing here. If there's any way I can help shape or test the conditioning interfaces as they come together, I'd be glad to. 🚀

**Comment by mglyn** [comment]
I agree with the direction in the latest discussion: keeping a parallel LTX-2.3 pipeline would now duplicate the shared conditioning/runtime architecture. Would it make sense to re-express this as a small integration on the current conditioning framework and validate it against the official LTX-2.5 keyframe semantics? If that is not planned, closing the current branch may avoid maintaining two implementations.
